### PR TITLE
MBS-11260 Fix redundant recalculations of dependencies in impact analysis

### PR DESCRIPTION
**Pull request by eugene-krivobokov**
It helps to decrease task graph calculation while configuration. 
On sample local build with enabled impact analysis: from 2m to 15s

### CODE REVIEW

File: subprojects/gradle/impact-shared/src/main/java/com/avito/impact/configuration/BaseConfiguration.kt [hunk 1]
```diff
@@ -46,19 +46,33 @@ abstract class BaseConfiguration(
     }
 
     fun allDependencies(includeSelf: Boolean = true): Set<BaseConfiguration> {
-        val dependencies = this.dependencies
-            .flatMap {
-                it.allDependencies(includeSelf = true)
```
Comment (eugene-krivobokov) [inline subprojects/gradle/impact-shared/src/main/java/com/avito/impact/configuration/BaseConfiguration.kt:51]: `allDependencies` calculated for each node recursively, but we need only for one node.

File: subprojects/gradle/impact-shared/src/main/java/com/avito/impact/configuration/BaseConfiguration.kt [hunk 2]
```diff
@@ -18,15 +18,15 @@ import java.io.File
  */
 abstract class BaseConfiguration(
     val module: InternalModule,
-    val types: Set<Class<out ConfigurationType>>
+    val type: Class<out ConfigurationType>
```
Comment (eugene-krivobokov) [inline subprojects/gradle/impact-shared/src/main/java/com/avito/impact/configuration/BaseConfiguration.kt:21]: Not used extensibility, don't need it right know.

File: subprojects/gradle/impact-shared/src/main/java/com/avito/impact/configuration/MainConfiguration.kt [hunk 3]
```diff
@@ -23,15 +21,6 @@ class MainConfiguration(module: InternalModule) : BaseConfiguration(
...
     }
 
-    override val hasChangedFiles: Boolean by lazy {
-        changedFiles()
-            .map { it.isNotEmpty() }
-            .onFailure {
-                project.logger.error("Can't find changes", it)
-            }
-            .getOrElse { true }
-    }
```
Comment (eugene-krivobokov) [inline subprojects/gradle/impact-shared/src/main/java/com/avito/impact/configuration/MainConfiguration.kt:33]: Duplicated inherited implementation

**Comment by dsvoronin** [comment]
oh shi~~

**Review by dsvoronin — APPROVED** [review]

**Pull request merged**